File: apps/mobile/lib/domain/models/order_fulfillment.dart
import 'models.dart';

class FulfillmentLine {
  final String productId;
  final int ordered,
      received,
      inTransit,
      remainingToDispatch,
      remainingToReceive;
  FulfillmentLine.fromJson(Json json)
    : productId = json['productId'],
      ordered = integer(json['ordered']),
      received = integer(json['received']),
      inTransit = integer(json['inTransit']),
      remainingToDispatch = integer(json['remainingToDispatch']),
      remainingToReceive = integer(json['remainingToReceive']);
  static int outstanding(StoreData data, String product) {
    if (data.raw.containsKey('outstandingSupply')) {
      return data
          .list('outstandingSupply')
          .where((line) => line['productId'] == product)
          .fold(0, (sum, line) => sum + integer(line['quantity']));
    }
    return data
        .list('orders')
        .expand((order) => objects(order['fulfillment']))
        .where((line) => line['productId'] == product)
        .fold(0, (sum, line) => sum + integer(line['remainingToReceive']));
  }
}

class OrderFulfillment {
  final String orderId, status;
  final int version;
  final List<FulfillmentLine> lines;
  OrderFulfillment.fromJson(Json json)
    : orderId = json['orderId'],
      status = json['status'],
      version = integer(json['version']),
      lines = List.unmodifiable(
        objects(json['lines']).map(FulfillmentLine.fromJson),
      );
}

File: apps/api/src/modules/operations/domain/order-fulfillment.ts
export interface FulfillmentLine {
  productId: string;
  ordered: number;
  cancelled: number;
  received: number;
  inTransit: number;
  remainingToDispatch: number;
  remainingToReceive: number;
}

export class OrderFulfillment {
  static calculate(
    lines: { productId: string; quantity: number }[],
    totals: { productId: string; received: number; inTransit: number }[],
    cancelledLines: { productId: string; quantity: number }[] = [],
  ): FulfillmentLine[] {
    const byProduct = new Map(totals.map((line) => [line.productId, line]));
    return lines.map((line) => {
      const { received = 0, inTransit = 0 } =
        byProduct.get(line.productId) ?? {};
      const cancelled =
        cancelledLines.find((c) => c.productId === line.productId)?.quantity ??
        0;
      return {
        productId: line.productId,
        ordered: line.quantity,
        received,
        cancelled,
        inTransit,
        remainingToDispatch: Math.max(
          0,
          line.quantity - received - inTransit - cancelled,
        ),
        remainingToReceive: Math.max(0, line.quantity - received - cancelled),
      };
    });
  }
}
